sys/linux: Additional Descriptions for Mali GPU Driver

Ji-Hyeon-Jeong opened the pull request:
I noticed that the syscall description for Mali GPU Driver fuzzing was missing the KBASE_IOCTL_KINSTR_PRFCNT_* related ioctls, so I added them.

tarasmadan commented:
Thank you! Let's run the tests first.

Ji-Hyeon-Jeong commented:
![image](https://github.com/user-attachments/assets/93c0b8a0-2de3-4aa0-81a3-4d2933fdf4a8)

In my environment, I was able to confirm that these syscall descriptions are working properly.

a-nogikh commented:
Please fix the following CI errors:
```
Error: Wrong commit subject format: 'sys/linux: Additional Descriptions for Mali GPU Driver'. Please use 'main/affected/package: short change description'. See docs/contributing.md for details. Regex pattern used to check commit subjects is '^(Revert "|(([a-z0-9/_.-]+|Makefile|CONTRIBUTORS|README.md)(, )?)+:\ [^A-Z].+[^.]$)'.
Error: Wrong commit subject format: 'sys/linux: Additional Descriptions for Mali GPU Driver'. Please use 'main/affected/package: short change description'. See docs/contributing.md for details. Regex pattern used to check commit subjects is '^(Revert "|(([a-z0-9/_.-]+|Makefile|CONTRIBUTORS|README.md)(, )?)+:\ [^A-Z].+[^.]$)'.
Error: Wrong commit subject format: 'Additional descriptions for Mali Driver'. Please use 'main/affected/package: short change description'. See docs/contributing.md for details. Regex pattern used to check commit subjects is '^(Revert "|(([a-z0-9/_.-]+|Makefile|CONTRIBUTORS|README.md)(, )?)+:\ [^A-Z].+[^.]$)'.
```

Ji-Hyeon-Jeong commented:
I'm sorry, I think I might have made a mistake somewhere. Could you please guide me on how to fix the CI? 😢

a-nogikh commented:
You just need to change your commit descriptions and force-push the branch, CI tests will run automatically

> [Merge branch 'master' into master](https://github.com/google/syzkaller/pull/5598/commits/d6a24c553dc107cd6642f184b3d97d0cba4ed023)

And that will cause tests to break again :( In our project, we rebase instead of doing merges. Please drop this commit.

Ji-Hyeon-Jeong commented:
Thank you so much for your guidance! I have fixed that problem. Could you please check again to confirm?

a-nogikh commented:
Now there's one more problem
```
sys/linux/dev_mali.txt:1:1: The file is not formatted/regenerated. Run 'make generate' and include it into the commit.
Error: The file is not formatted/regenerated. Run 'make generate' and include it into the commit.
```

Also please notice that now you have 3 commits with identical descriptions:
https://github.com/google/syzkaller/pull/5598/commits

Ji-Hyeon-Jeong commented:
Thank you so much! I have reviewed and fixed the issues you pointed out. Could you kindly check again and let me know if everything looks good?

ramosian-glider commented:
Please squash "run make extract" into "Additional descriptions for Mali Driver", there's no need to have intermediate patches in the tree.

Ji-Hyeon-Jeong commented:
Thank you for your advice. I made the changes as requested.

ramosian-glider commented:
@Ji-Hyeon-Jeong, where do these constants actually come from? Is there any source that we can reference here?

Ji-Hyeon-Jeong commented:
I've extracted all constant from ARM Mali5th Gen Driver r53p0 version, (https://developer.arm.com/downloads/-/mali-drivers/5th-gen-gpu-architecture-kernel).

Every constants I've added come from `driver/product/kernel/include/uapi/gpu/arm/midgard/mali_kbase_hwcnt_reader.h`(`457L~463L & 495L~504L`). Sorry for the delay in checking this 😢 — I had something else to take care of yesterday.

![image](https://github.com/user-attachments/assets/02c9ff5d-8df5-4111-a91b-08bec93c752b)

ramosian-glider commented on sys/linux/dev_mali.txt:
```diff
@@ -331,6 +335,17 @@ kbase_ioctl_context_priority_check {
 }
 
+prfcnt_control_cmd {
+	cmd		flags[control_cmd, int8]
```
This must be int16:

```
struct prfcnt_control_cmd {
        __u16 cmd;
        __u16 pad[3];
        __u64 user_data;
};
```

Ji-Hyeon-Jeong replied:
I have addressed the points you mentioned! Thank you so much for your valuable feedback!

ramosian-glider commented on sys/linux/dev_mali.txt:
```diff
@@ -39,13 +40,16 @@ ioctl$KBASE_IOCTL_CS_QUEUE_KICK(fd fd_bifrost, cmd const[KBASE_IOCTL_CS_QUEUE_KI
 ioctl$KBASE_HWCNT_READER_PUT_BUFFER_WITH_CYCLES(fd fd_hwcnt, cmd const[KBASE_HWCNT_READER_PUT_BUFFER_WITH_CYCLES], arg ptr[in, kbase_hwcnt_reader_metadata_with_cycles])
 ioctl$KBASE_HWCNT_READER_GET_API_VERSION_WITH_FEATURES(fd fd_hwcnt, cmd const[KBASE_HWCNT_READER_GET_API_VERSION], arg ptr[out, kbase_hwcnt_reader_api_version])
+ioctl$KBASE_IOCTL_KINSTR_PRFCNT_CMD(fd fd_kinstr, cmd const[KBASE_IOCTL_KINSTR_PRFCNT_CMD], arg ptr[in, prfcnt_control_cmd])
+ioctl$KBASE_IOCTL_KINSTR_PRFCNT_GET_SAMPLE(fd fd_kinstr, cmd const[KBASE_IOCTL_KINSTR_PRFCNT_GET_SAMPLE], arg ptr[in, prfcnt_sample_access])
```
Should be `arg ptr[out, ...]`

Ji-Hyeon-Jeong replied:
I have addressed the points you mentioned! Thank you so much for your valuable feedback!

ramosian-glider commented on sys/linux/dev_mali.txt:
```diff
@@ -331,6 +335,17 @@ kbase_ioctl_context_priority_check {
+prfcnt_control_cmd {
+	cmd		flags[control_cmd, int8]
+	padding		array[const[0, int16], 3]
+	user_data	ptr64[in, array[int64]]
```
Do we have insight into the size of this array and how it's being used?

Ji-Hyeon-Jeong replied:
```
/** struct prfcnt_control_cmd - Control command
 * @cmd:       Control command for the session.
 * @pad:       Padding bytes.
 * @user_data: Pointer to user data, which will be returned as part of
 *             sample metadata. It only affects a single sample if used
 *             with CMD_SAMPLE_SYNC or CMD_SAMPLE_ASYNC. It affects all
 *             samples between CMD_START and CMD_STOP if used with the
 *             periodic sampling.
 */
struct prfcnt_control_cmd {
	__u16 cmd;
	__u16 pad[3];
	__u64 user_data;
};
```

As seen in the definition of prfcnt_control_cmd, the user_data field is described as a "pointer to user data." However, this description alone is not sufficient to clearly determine the state of the array. I confirmed that user_data is used in functions such as kbasep_kinstr_prfcnt_client_output_sample to store certain data, but I was unable to gather additional information to definitively conclude its exact state.

I updated the definition to ptr64[out, array[int64]] (changing it from the original ptr64[in, array[int64]]) as I couldn’t think of a more appropriate way to define it. If this definition needs to be revised, I would greatly appreciate your suggestions on how it can be improved. Thank you so much for your valuable advice!

ramosian-glider replied:
I see, thanks. I think we're good for now.
